llvmlite/ir/values.py

```diff
@@ -980,9 +984,10 @@ def descr_prototype(self, buf):
         cconv = self.calling_convention
         prefix = " ".join(str(x) for x in [state, linkage, cconv, ret] if x)
         metadata = self._stringify_metadata()
-        pt_str = "{prefix} {name}({args}{vararg}) {attrs}{metadata}\n"
+        section = 'section "{}"'.format(self.section) if self.section else ''
+        pt_str = "{prefix} {name}({args}{vararg}) {section}{attrs}{metadata}\n"
```

Comment:
Yes, you are right. The attributes should come before the section. I didn't read the langref docs carefully enough.

About the `{attrs}{metadata}` bug, I'm guessing it is because if there is a space between (`{section} {attrs} {metadata}`) there would be two spaces (or even three) if no `{section}` or `{attrs}` were specified, but `{metadata}` was.
```llvm
declare i32 @"some_function"(i32 %".1", i32 %".2", double %".3", i32* %".4")   !dbg !0
```
This would cause the tests to fail as they compare an exact string with only one space before metadata. On the other hand, specifying more than one would also lead to test failure (e.g., if there are tests having both `{attrs}` and `{metadata}`). 

Either the tests should be updated to accept more than one space (if that is allowed in the langref docs, I can't really tell) or we should make sure that a space is added only if the relevant portion (`{section}`, `{attrs}` or `{metadata}`) is used so that there is only one space between the different parts of the declaration.

I will look into both issues and update the pull request.

Reply:
I think your assessment regarding additional spaces in the above is accurate. IIRC additional space in LLVM IR is fine, however, producing LLVM IR with regularised formatting is a good aim and it helps with unit testing, so if it's possible to fix this then I think it should be fixed! Many thanks for volunteering to look at this as part of this PR.

Reply:
In the original code
```python
pt_str = "{prefix} {name}({args}{vararg}) {attrs}{metadata}\n"
```
we have four cases:
No `{attrs}` or `{metadata}` gives a space at the end before `\n`
```llvm
declare i32 @"some_function"(i32 %".1", i32 %".2", double %".3", i32* %".4") \n
```
Only `{attrs}` or `{metadata}` gives no space before `\n`
```llvm
declare i32 @"some_function"(i32 %".1", i32 %".2", double %".3", i32* %".4") alwaysinline\n
```
```llvm
declare i32 @"some_function"(i32 %".1", i32 %".2", double %".3", i32* %".4") !dbg !0\n
```
And the erroneous case you noted
```llvm
declare i32 @"some_function"(i32 %".1", i32 %".2", double %".3", i32* %".4") alwaysinline!dbg !0\n
```
This makes it hard to test as you need to know that you get a space before `\n` in one case and in another there is no such space. This will be even harder if we add space between `{attrs}`, `{section}` and `{metadata}`
```python
pt_str = "{prefix} {name}({args}{vararg}) {attrs} {section} {metadata}\n"
```
Is it specified somewhere that there should always be just one space between different parts of the prototype?
Otherwise I would reduce multiple spaces to one and always ensure that there is no space before `\n`. That would make it consistent.
Any thoughts?

Reply:
I don't think there's a specification for the amount of space between the parts, @sklam do you recall?

> Otherwise I would reduce multiple spaces to one and always ensure that there is no space before \n. That would make it consistent.


I think it'd be good to just do this as it's predictable and consistent. I suppose each of the `{attr}`, `{section}` and `{metadata}` need to return empty string or a valid string with a single space on the end, think that gets the consistent output, is that similar to what you are thinking?

Reply:
>I suppose each of the {attr}, {section} and {metadata} need to return empty string or a valid string with a single space on the end, think that gets the consistent output, is that similar to what you are thinking?

Yes, either that or just reduce multiple spaces to one with `re.sub`.

I will also update the tests to comply with this and maybe add some new tests to test (missing) combinations.

Reply:
That would be great, many thanks for your help with this!

Reply:
> I don't think there's a specification for the amount of space between the parts, @sklam do you recall?

Don't recall there's any specific spec around whitespaces. Just the normal of having at least a whitespace to help separate out tokens when there are no separators.